src/devices/machine/swtpc8212.cpp

```diff
@@ -0,0 +1,605 @@
+
+		for (int bit = 0; bit < 8; bit++)
+		{
+			int dout = BIT(data, 7);
+			uint32_t font_color = 0;
+			if ((dcursor ^ dout) && de)
+			{
+				if (intensity)
+					font_color = rgb_t(0x10, 0xff, 0x10);
+				else
+					font_color = rgb_t(0x00, 0xd0, 0x00);
+			}
+			bitmap.pix32(y, x++) = font_color;
+			data <<= 1;
+		}
+
+		// Gap between characters
+		if (BIT(m_latch_data, 7) == 0)
+			x++;
+	}
+}
+
+void swtpc8212_device::device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr)
+{
+	switch (id)
+	{
+	case BELL_TIMER_ID:
+		m_beeper->set_state(0);
+		break;
+	}
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::rs232_conn_dcd_w)
+{
+	m_uart->dcd_w(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::rs232_conn_dsr_w)
+{
+	m_uart->dsr_w(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::rs232_conn_ri_w)
+{
+	m_uart->ri_w(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::rs232_conn_cts_w)
+{
+	m_uart->cts_w(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::rs232_conn_rxd_w)
+{
+	m_uart->rx_w(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::write_txd)
+{
+	m_rs232_conn_txd_handler(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::write_dtr)
+{
+	m_rs232_conn_dtr_handler(state);
+}
+
+WRITE_LINE_MEMBER(swtpc8212_device::write_rts)
+{
+	m_rs232_conn_rts_handler(state);
+}
+
+void swtpc8212_device::device_resolve_objects()
+{
+	m_rs232_conn_dtr_handler.resolve_safe();
+	m_rs232_conn_rts_handler.resolve_safe();
+	m_rs232_conn_txd_handler.resolve_safe();
+}
+
+void swtpc8212_device::device_start()
+{
+	m_bell_timer = timer_alloc(BELL_TIMER_ID);
+
+	save_item(NAME(m_latch_data));
+	save_item(NAME(m_keyboard_data));
+	save_item(NAME(m_printer_data));
+	save_item(NAME(m_printer_data_ready));
+}
+
+void swtpc8212_device::device_reset()
+{
+	m_keyboard_data = 0;
+	m_pia0->cb1_w(ASSERT_LINE);
+	m_keypad_data = 0;
+	m_pia0->ca1_w(ASSERT_LINE);
+
+	m_latch_data = 0x1f;
+
+	m_beeper->set_state(0);
+
+	m_printer_data = 0;
+	m_printer_data_ready = 1;
+	m_pia1->ca1_w(CLEAR_LINE);
+
+	if (m_one_stop_bit->read())
+	{
+		// Patch the firmware to use one stop bit.
+		uint8_t* program = memregion("program")->base();
+		program[0x01ad] = 0x02;
+	}
+
+	if (m_keypad_mod->read())
+	{
+		// Patch the firmware to modify the keypad key codes.
+		uint8_t* program = memregion("program")->base();
+		program[0x00e5] = 0x00; // Form, was Ctrl-L, now unused
+		program[0x00e6] = 0x00; // Xmit, was Ctrl-W, now unused
+		program[0x00e8] = 0x06; // Right, was Ctrl-I, now Ctrl-F
+		program[0x00e9] = 0x02; // Left, was Crtl-D, now Ctrl-B
+		program[0x00ea] = 0x01; // Home, was Crtl-P, now Ctrl-A
+		program[0x00ec] = 0x11; // Insert, was Crtl-Y, now Ctrl-Q
+		program[0x00ed] = 0x16; // Scroll Up, was Crtl-N, Ctrl-V
+		program[0x00ee] = 0x10; // Up, was Crtl-A, now Ctrl-P
+		program[0x00f0] = 0x0b; // Delete, was Crtl-Z, now Ctrl-K
+		program[0x00f1] = 0x00; // Scroll Down, was Crtl-O, now unused
+		program[0x00f2] = 0x0e; // Down, was Crtl-B, now Ctrl-N
+	}
```

Comment:
I really don't like these kinds of patches.  It's not the MAME way.  Can you please do this using the cheat plugin, or some other way using Lua scripting?  (Yes, this is another reminder that we need a kind of "IPS for multiple ROM regions" format to support soft-patching via a plugin to allow ROM-only hacks to run in mainline MAME easily.)

Reply:
Yes, that had occurred to me. I am happy to pull that for now so people can consider it further if this is blocking this PR?

Here's my perspective. Systems in the late 1970's and early 1980's were limited, the ROMs and EPROMs were limited, there was a large DIY element to getting a computer running and fitting components together. Documentation commonly came with hardware mods to implement various functionality. Since the ROMs were limit, their functionality was limited, and it was not unexpected to have to patch a ROM here or there, it was not considered 'cheating', and even working with these systems today that is how it rolls.

If I had a 8212 terminal and wanted to connect it to a UniFlex SWTPC09 today then I would be patching the ROM to modify the codes send by the cursor keys - that's not a emulator hack, rather it's just how it rolls with this tech.

Here's another example. The CT-82 version 1 ROM supports 9600 baud, 1 stop bit, no parity, and that's it. Version 2 re-purposed the dips switches and jumpers to add a range of other formats, but all with 2 stop bits, there is no support for 1 stop bit and that breaks compatibility with many systems that this terminal was used with in the day. To get everything working together with this vintage tech does need some ROM patching, it's just how it rolled.

So given that, how to fit in with the 'MAME approach'?

1. Just use some representative ROM, potentially leaving the MAME emulators unworkable in practice. Would the MAME project even want required modifications noted in the source code or has that all become too much and are you wanting to separate the hardware emulation from the documentation of the software history and how to get things working together?

2. Multiple ROMs, for each combination of patches. This would appear to fit with the approach taken by some other machines in MAME. However where there are multiple patches there is an explosion of ROM combinations which is not great.

3. A run time loadable ROM, so people supply their own.

Approaches 2 and 3 leave the details of the patch to be explained elsewhere, rather than in the MAME emulator source code, and I was hoping to dump what knowledge I can in these emulators, I can't host a website anymore, people leave unfortunately.

4. ROM patches in the MAME emulator for some key trivial changes. This avoids having ROMs for each combination of patches. It gives some direction as to what changes have been made and where and why. This is the approach I have taken for these vintage systems.